pkg/resources/provisioning.cattle.io/v1/cluster/validator.go

```diff
@@ -103,6 +113,14 @@ func (p *provisioningAdmitter) Admit(request *admission.Request) (*admissionv1.A
 		if err := p.validateCloudCredentialAccess(request, response, oldCluster, cluster); err != nil || response.Result != nil {
 			return response, err
 		}
+
+		if response = p.validateAgentEnvVars(request, oldCluster, cluster); response.Result != nil {
```

Comment:
I would handle the "ok" cases differently than this. Right now this depends on `response.Result` being nil for `ResponseAllowed()` which I think is very dangerous long term. I would recommend one of the following options:
- Only return a response if it fails validation and return nil otherwise. In this function, check `response != nil` to see if you should break with a response
- In this function, check if `!response.Allowed` - this way, if it fails we short-circuit and return early.
- Return a bool and an `*admissionv1.AdmissionResponse`. Use the bool to check if the response was allowed or not. 

Big thing here is not to closely couple this code to what the allowed response does like it's currently doing.